controllers/controllers/cluster_controller.go

```diff
@@ -32,32 +38,70 @@ const (
 
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
-	client    client.Client
-	log       logr.Logger
-	validator *vsphere.Validator
-	defaulter *vsphere.Defaulter
-	tracker   *remote.ClusterCacheTracker
+	client                  client.Client
+	log                     logr.Logger
+	validator               *vsphere.Validator
+	defaulter               *vsphere.Defaulter
+	tracker                 *remote.ClusterCacheTracker
+	buildProviderReconciler BuildProviderReconciler
 }
 
-func NewClusterReconciler(client client.Client, log logr.Logger, scheme *runtime.Scheme, govc *executables.Govc, tracker *remote.ClusterCacheTracker) *ClusterReconciler {
+// TODO: this is not ideal and will need a refactor. I will follow up but for now this
+// allows us to decouple the cluster reconciler main logic from provider specific logic
+type BuildProviderReconciler func(datacenterKind string, client client.Client, log logr.Logger, validator *vsphere.Validator, defaulter *vsphere.Defaulter, tracker *remote.ClusterCacheTracker) (clusters.ProviderClusterReconciler, error)
+
+func NewClusterReconciler(client client.Client, log logr.Logger, scheme *runtime.Scheme, govc *executables.Govc, tracker *remote.ClusterCacheTracker, buildProviderReconciler BuildProviderReconciler) *ClusterReconciler {
```

Comment:
will arg govc be removed? will validator and defaulter be part of the `buildProviderReconciler`? I don't quite follow why this decouples the main logic from provider logic